nemo/collections/asr/parts/k2/topologies.py

```diff
@@ -139,3 +146,41 @@ def build_minimal_topo(tokens: List[int]) -> 'k2.Fsa':
     ans = k2.Fsa.from_str(arcs, num_aux_labels=1)
     ans = k2.arc_sort(ans)
     return ans
+
+
+class RnntEmissionAdapterBuilder(object):
+    """TBD
+    """
+
+    def __init__(self, tokens: List[int]):
+        assert -1 not in tokens, "We assume -1 is ID of the final transition"
+        assert 0 in tokens, "We assume 0 is ID of the epsilon symbol"
+        assert 1 in tokens, "We assume 1 is ID of the blank symbol"
+
+        self.tokens = tokens
+
+    def __call__(self, adapter_lengths: Union[torch.Tensor, List[int]]) -> 'k2.Fsa':
+        # if you don't make adapter_lengths a list beforehand,
+        # "i" will be implicitly converted to int, and this will always be considered a cache miss
+        return k2.create_fsa_vec([self._build_single_adapter(i) for i in adapter_lengths.tolist()])
+
+    @lru_cache(maxsize=1024)
+    def _build_single_adapter(self, adapter_length: int) -> 'k2.Fsa':
+        assert adapter_length >= 1, "`adapter_length` cannot be less than one"
+
+        num_states = len(self.tokens)
+        blank_num = 1
+        first_eps_state = adapter_length + 1
+        final_state = adapter_length * 2 + 1
+        arcs = ""
+        for i in range(adapter_length):
```

Comment:
It seems that you can significantly speedup this by using `k2.Fsa(arcs)` constructor, where `arcs` is a tensor with the size `(num_arcs, 4)` of type `torch.int32`. But it will less readable(

Reply:
Thanks for the suggestion!
These adapters are cached pretty fast (in practice, there are <120 calls most of the time), so I'd rather keep the implemetation readable.